%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.html and the instructions for   %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

%\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
\documentclass[parskip=full]{bmcart} %linenumbers
\setlength{\parskip}{10pt}
\usepackage{array,multirow,graphicx}
\usepackage{lineno} % Lindsay added
\usepackage[lined]{algorithm2e}
%%% Load packages
%\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
%\RequirePackage[authoryear]{natbib}% uncomment this for author-year bibliography
%\RequirePackage{hyperref}
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\usepackage{makecell}
\renewcommand\theadalign{bc}
\renewcommand\theadfont{\bfseries}
\renewcommand\theadgape{\Gape[4pt]}
\renewcommand\cellgape{\Gape[4pt]}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\def\includegraphic{}
\def\includegraphics{}



%%% Put your definitions there:
\startlocaldefs
\endlocaldefs


%%% Begin ...
\begin{document}
\linenumbers

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Software}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{bigPint: A Bioconductor package that makes big data pint-sized}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Specify information, if available,       %%
%% in the form:                             %%
%%   <key>={<id1>,<id2>}                    %%
%%   <key>=                                 %%
%% Comment or delete the keys which are     %%
%% not used. Repeat \author command as much %%
%% as required.                             %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\author[
addressref={aff1},                   % id's of addresses, e.g. {aff1,aff2}
corref={aff1},                       % id of corresponding address, if any
email={lindsayannerutter@gmail.com}   % email address
]{\inits{LR}\fnm{Lindsay} \snm{Rutter}}
\author[
addressref={aff2},
email={dicook@monash.edu}
]{\inits{DC}\fnm{Dianne} \snm{Cook}}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%% Repeat \address commands as much as      %%
%% required.                                %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address[id=aff1]{
\orgname{Bioinformatics and Computational Biology Program, Iowa State University},
\city{Ames},
\cny{USA}
}
\address[id=aff2]{
\orgname{Econometrics and Business Statistics, Monash University},
\city{Clayton VIC},
\cny{Australia}
}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter short notes here                   %%
%%                                          %%
%% Short notes will be after addresses      %%
%% on first page.                           %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{artnotes}
%\note{Sample of title note}     % note to the article
%\note[id=n1]{Equal contributor} % note, connected to author
\end{artnotes}

\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% The Abstract should not exceed 100 words. Please minimize the use of abbreviations and do not cite references in the abstract. The abstract should be unstructured.

\begin{abstractbox}

\begin{abstract} % abstract
We developed bigPint, an interactive data visualization package available on Bioconductor. Our software introduces new visualization technology that enables independent layers of interactivity, aiding in the exploration of large datasets. The ability to select and aggregate data, link between plots, and tailor aesthetics in intelligent ways are all useful features of bigPint. Researchers can analyze and present their increasingly large biological datasets using intuitive and reproducible plots from bigPint. Developers can leverage our open-source code to develop additional interactive visualization tools for computational biology tasks.

\end{abstract}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The keywords begin here (3-10)           %%
%%                                          %%
%% Put each keyword in separate \kwd{}.     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{keyword}
\kwd{interactive}
\kwd{visualization}
\kwd{RNA-seq}
\kwd{Bioconductor}
\kwd{clustering}
\kwd{Shiny}
\kwd{Plotly}
\kwd{htmlwidgets}
\kwd{ggplot2}
\kwd{R}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
%\end{fmbox}% uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%% start of article main body
% <put your article body there>

%%%%%%%%%%%%%%%%
%% Background %%
%%

%\cite{koon,oreg,khar,zvai,xjon,schn,pond,smith,marg,hunn,advi,koha,mouse}

% aims, summary of literature, why study important
\section*{Background}

Interactive data visualization is increasingly imperative in the biological sciences \cite{o2010visualizing}. When performing RNA-seq studies, researchers wish to determine which genes are differentially expressed between treatment groups. Interactive visualization can help them assess differentially expressed gene (DEG) calls before performing any subsequent functional enrichment analyses. New visualization tools for genomic data have incorporated interactive capabilities, and some believe this trend could enhance the exploration of genomic data in the future \cite{pavlopoulos2015visualizing}. Despite the growing appreciation of the inherent value of interactive graphics, the availability of effective and easy-to-use interactive visualization tools for RNA-seq data remains limited.

Interactive visualization tools for genomic data can have restricted access when only available on certain operating systems and/or when requiring payment \cite{seo2002interactively, ahlberg1996spotfire, chu2001genespringtm}. These limitations can be removed when tools are published on open-source repositories. Indeed, the Bioconductor project aims to foster interdisciplinary scientific research by promoting transparency and reproducibility while allowing software content to be used on Windows, MacOS, and Linux \cite{gentleman2004bioconductor}. Bioconductor software is written in the R programming language, which also provides statistical and visualization methods that can facilitate the development of robust graphical tools. Several interactive visualization methods for genomic data have been developed using Shiny, which is also based on the R programming language \cite{rue2018isee, schultheis2018wilson, hughes2017expressiondb}. 

We recently developed bigPint, an interactive data visualization software package available on Bioconductor. The bigPint package allows users to visually explore many types of large multivariate datasets, even though it was more specifically developed for RNA-seq data. In a recent methods paper, we used public RNA-seq datasets to demonstrate how bigPint graphics can help biologists detect crucial issues with normalization methods and DEG designation in ways not possible with numerical models \cite{rutter1}. We also applied bigPint visualization tools in a recent research paper that sought to elicit how nutrition and viral infection affect the honey bee transcriptome \cite{rutter2}. In the current paper, we will now explain the technical innovations and merits of the bigPint package, including new interactive visualization techniques that we believe can be helpful in the development and usage of future biological visualization software. The bigPint website is available at https://lindsayrutter.github.io/bigPint and contains short vignette articles that provide example analysis pipelines, all written in reproducible code.

\section*{Results}

%%%%%%%%%%%%% Basic usage %%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Basic input}

Each method in bigPint requires an input parameter data object. If a researcher is using the package to visualize RNA-seq data, then this data object should be a count table that contains the read counts for all genes of interest. The value in row \textit{i} and column \textit{j} should indicate how many reads have been assigned to gene \textit{i} in sample \textit{j}. This is the same input format required in popular RNA-seq count-based statistical packages, such as DESeq2, edgeR, limma, EBSeq, and BaySeq \cite{deseq2, edger, limma, ebseq, bayseq}. 

Several methods in bigPint also require an input parameter dataMetrics object. If a researcher is using the package to visualize RNA-seq data, then this dataMetrics object should be a subset of the data (usually DEGs) where each case includes quantitative values of interest (such as fold change and FDR). This information can be easily derived from popular RNA-seq numerical analysis packages. Again, this framework allows users to work smoothly between visualizations in the bigPint package and models in other Bioconductor packages, complying with the belief that the most efficient way to analyze large datasets is to iterate between models and visualizations.

%%%%%%%%%% Original features %%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Original features}

\subsubsection*{1. Independent layers of interactivity}

The Bioconductor community advanced the boundaries of biological visualization in the past and generally believes that modern interactive technology must be incorporated to continue these advancements \cite{gentleman2004bioconductor}. We will define the term \textit{geom-drawing interactivity} to indicate user queries that draw geoms (graphical representations of the data, such as lines, hexagons, and points). This could mean the user adjusts sliders or selects buttons to draw a subset of the data from the database as geoms (such as points). We will define the term \textit{geom-manipulating interactivity} to indicate user queries that alter already-drawn geoms. This could mean the user hovers over a geom (such as a hexagon) and obtains its associated metadata (such as the names of its contained genes). It could also mean the user zooms and pans to further alter how already-drawn geoms are displayed.

Our package introduces what we believe is a fairly new interactive visualization technology that is useful in the exploration of large biological datasets. Our technique allows for two independent layers of interactivity, for the foreground and background of the plot respectively. Each layer can include both \textit{geom-drawing} and \textit{geom-manipulating} interactivity. Our new technology can enhance the exploration of large datasets, especially in cases where one layer contains large amounts of data (such as the full dataset) and the other layer contains smaller amounts of data (such as a data subset). Because the layers are independent, users can save time and computation by keeping the layer with more data unaltered while only redrawing the layer with less data.

We achieved our independent double-layered interactivity using the onRender() method of the htmlwidgets package \cite{htmlwidgets}. This method had the potential for the foreground layer to be overlaid via plotly traces while the plotly background layer did not need to be redrawn, something that could not foreseeably be achieved with the native onRender() method of the plotly package \cite{plotly}. Specifically, the htmlwidgets onRender() method contains three input parameters: an HTML Widget object, a character vector containing JavaScript code, and a list of R objects that can be serialized to JSON format. To develop our technique, we specified a plotly object as the HTML Widget object, which allowed for an interactive background. Within the method, we wrote JavaScript code that enabled interactive foregrounds to be updated without redrawing interactive backgrounds. We used the R object list to transfer count tables and DEG lists into the method. In some our our applications, users can link between the layers of different interactive plots. This functionality was achieved by sending custom messages between the Shiny software and the JavaScript code within the htmlwidgets method \cite{shiny}. We provide pseudocode and documented code for readers who wish to understand the details of how we created our interactive software (see Table~\ref{table:table2}). We will now briefly explain how our two-layered interactivity method can improve upon several of the RNA-seq visualization tools in our package. 

\subsubsection*{1a. Scatterplot matrices}

Scatterplot matrices have appeared in statistical graphics literature for almost four decades and used across various fields of multivariate research \cite{becker1984brushing, carr1984graphical, tufte2001visual, tukey1981graphical}. Previous user studies have shown that participants performed better when using animated rather than static versions of scatterplot matrices. Users also preferred animated scatterplot matrices and found them easier to understand as they can alleviate overplotting issues \cite{chen2018using}. Rendering scatterplot matrices interactive is promising but challenging with large datasets \cite{carr1987scatterplot}. The number of background geoms that need to be drawn grows exponentially by dimension size: $n$-dimensional data corresponds to $n^2$ scatterplots. Our two-layered interactive visualization technology improves upon this dilemma by allowing details of interest to be superimposed in the foreground while the massive number of geoms in the background does not require redrawing. See Tables~\ref{table:table1} and~\ref{table:table2} for details (video, pseudocode, code, and application link) about our interactive scatterplot matrices.

\subsubsection*{1b. Litre plots}

Problems still remain when scatterplot matrices are applied to large datasets. Physical space requirements increase exponentially. Hence, when extended to large dimensions, it becomes difficult to mentally link many small plots within the matrix \cite{kerren2007human}. Several techniques have been proposed to ameliorate this problem. Three dimensional scatterplots are useful but can cause occlusion and depth perception issues \cite{kerren2007human}. Other techniques like grand tours \cite{asimov1985grand}, projection pursuits \cite{friedman1974projection, cook1995grand}, and scagnostics \cite{wilkinson2005graph} have been proposed.

Even though these alternative techniques are useful, they may not simultaneously display distributions across all cases (genes) and variables (samples). We generally want to compare replicate and treatment variability in RNA-seq data, which can be visually accomplished by plotting all genes and samples. We also want to superimpose DEGs to determine how their read count variability compares to that of the whole dataset. In light of this, we developed a plot that collapses the scatterplot matrix onto one Cartesian coordinate system, allowing users to visualize all read counts from one DEG of interest onto all read counts of all genes in the dataset. We call this new plot a repLIcate TREatment (``litre'') plot. An in depth explanation about the litre plot can be found in our previous methods paper \cite{rutter1}.

We believe our two-layered interactive visualization method is an indispensable component of the litre plot. Drawing the background (all genes in the dataset) is the time-limiting step, whereas drawing the foreground (one DEG of interest) is immediate. Most users would like to superimpose DEGs from a list one by one onto the background. This process would be unnecessarily time-prohibiting if the background needed to be redrawn each time the user progressed to the next DEG. Fortunately, our technology allows the user to immediately redraw the interactive foreground (the DEG of interest) while the background (all genes in the data) remains unchanged but preserved in its interactive capabilities. See Tables~\ref{table:table1} and~\ref{table:table2} for details (video, pseudocode, code, and application link) about our interactive litre plots.

\subsubsection*{1c. Volcano plots}

Volcano plots draw significance and fold change on the vertical and horizontal axes respectively. In RNA-seq studies, volcano plots allow users to check that genes were not falsely deemed significant due to outliers, low expression levels, and batch effects \cite{li2011application}. Researchers benefit from the ability to quickly identify individual gene names in the volcano plot. This was previously achieved with the identify() method in R, which identifies the closest point in a scatterplot to the position nearest the mouse click \cite{li2011application}. The interactive volcano plot in bigPint can identify individual gene names in a less ambiguous fashion by responding to users hovering \textit{directly} over corresponding points. It also improves upon traditional volcano plots by allowing users to threshold on statistical values in order to immediately update the superimposed gene subset without having to redraw the more computationally-heavy background that contains all genes. See Tables~\ref{table:table1} and~\ref{table:table2} for details (video, pseudocode, code, and application link) about our interactive volcano plots.

\subsubsection*{2. Consecutive box selection}

The bigPint package provides interactive tools for consecutive box selection. A box selection is a rectangular query drawn directly on a two-dimensional graph. Users can specify a box selection by clicking on the desired starting point of the rectangular query and dragging the mouse pointer to the desired opposite corner point of the rectangular query. This procedure for generating rectangles is widely used in interactive programs and should be familiar to most users \cite{hochheiser2003dynamic}. After the user releases the mouse, the query is processed and only the data cases that were inside the specified rectangle remain. More precisely, a data case remains in a box selection queried between ($x_1$, $y_1$) and ($x_2$, $y_2$) if every point within $x_1 \leq x \leq x_2$ is also within $y_1 \leq y \leq y_2$ (where $y_2 \geq y_1$ and $x_2 \geq x_1$). The user can specify consecutive queries with multiple box selections. The consecutive box selection model is convenient in cases where identical thresholds are desired over adjacent features. In these cases, a single box selection of width $w$ can be used to simultaneously query the same threshold across $w$ features. This process is an improvement over single-feature box selection widgets, where $w$ individual queries would be required \cite{hochheiser2003dynamic}. 

Consecutive box selection may have originally been designed for time series data, but has since proven useful for detecting patterns in gene expression data. Combined with parallel coordinate plots, the consecutive box selection technique has been used to elicit candidate regulatory splice sequences showing high values at some positions and low values at other positions \cite{hochheiser2003dynamic}. In RNA-seq, this technology can also be used to investigate differential expression showing high read counts for one treatment group and low read counts for another treatment group, requiring a consecutive query. Consecutive box selection tools have been published for gene expression analysis sofware that was restricted for certain operating systems \cite{hochheiser2003dynamic}. We believe that publishing consecutive box selection tools in a platform like R can be useful for computational biologists using various operating systems. See Tables~\ref{table:table2} for details (video, pseudocode, code, and application link) about our interactive parallel coordinate plots that feature consecutive box selection.

%%%%%%%%%%% Useful features %%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Useful features}

\subsubsection*{Tailoring and saving static plots}

Static plots can be saved as list objects in the R workspace and/or as JPG files to a directory chosen by the user. Saving plots into the R workspace allows users to integrate them into analysis workflows. It also allows them to tailor the plots (such as adding titles and changing label sizes) using the grammar of graphics via the conventional + syntax. Saving plots to a directory allows users to keep professional-looking files that can be inserted into proposals and talks. By default, the bigPint package saves static plots both in the R workspace and a directory (the default location is tempdir()).

\subsubsection*{Second feature layer}

Both static and interactive plots allow for a subset of data to be plotted in a different manner than the full dataset. When analyzing RNA-seq data, this second feature layer could represent DEGs. There are three options for creating data subsets with static plots. First, users can threshold the previously-mentioned dataMetrics object by one of its quantitative variables. Second, users can simply declare a geneList object that contains the list of data subset IDs. Third, the user can simply leave the dataMetrics and geneList objects to their default value of NULL and not overlay any data subsets.

\subsubsection*{Group comparison filters}

When users create static plots, the package automatically creates a separate plot for each pairwise combination of treatment groups from the inputted data. When users explore interactive plots, fields are dynamically generated from the inputted data so that any pairwise combination of treatment groups can be selected by buttons. Users can then quickly flip between contrasts in their data. The bigPint package comes with an example soybean cotyledon dataset that has three treatment groups, which is used across several easy-to-follow articles on the package website. These assets can assist users who have data containing more than two treatment groups. 

\subsubsection*{Hexagonal binning}

Most bigPint plots represent genes using point geoms (where each point represents one gene) or hexagonal binning geoms (where each hexagon color represents the number of genes in that area). Plotting each gene as a point allows for ideal levels of detail but overplotting can occur as the data increases, which makes it difficult to determine how many genes are in a given area. Hexagonal binning has been used in prior software to successfully manage overplotting issues \cite{carr1987scatterplot, harshbarger2017deiva} and has shown superior time performance because less geom objects need to be plotted. The bigPint package allows users to draw the background using either geom, as preferences can depend on the dataset.

\subsubsection*{Hierarchical clustering}

Users can conduct hierarchical clustering analyses on their data using the function plotClusters(). By default, the resulting clusters will be plotted as parallel coordinate lines superimposed onto side-by-side boxplots that represent the five-number summary of the full dataset. There are three main approaches in the plotClusters() function:

\begin{itemize}  
\item Approach 1: The clusters are formed by clustering only on a user-defined subset of data (such as significant genes). Only these user-defined genes are overlaid as parallel coordinate lines. 
\item Approach 2: The clusters are formed by clustering the full dataset. Then, only a user-defined subset of data (such as significant genes) are overlaid as parallel coordinate lines.
\item Approach 3: The clusters are formed by clustering the full dataset. All genes are overlaid as parallel coordinate lines.
\end{itemize}

The clustering algorithm is based on the hclust() and cutree() functions in the R stats package. It offers the same set of agglomeration methods (``ward.D'', ``ward.D2'', ``single'', ``complete'', ``average'', ``mcquitty'', ``median'', and ``centroid'') with ``ward.D'' as the default. In many cases, users may want to save clusters derived from the plotClusters() function for later use, such as to overlay them onto scatterplot matrices, litre plots, and volcano plots. The gene IDs of each cluster can be saved as .RDS files for this purpose by setting the verbose option of the plotClusters() function to a value of TRUE.

\subsubsection*{Various plot aesthetics}

Users can modify various aesthetics for both static and interactive plots, including geom size. Some plots also provide alpha blending, which can benefit users plotting large datasets as parallel coordinate lines \cite{unwin2007computational}. Statistical coloring is inconsistent in numerous packages even though it can greatly enhance biological data visualization \cite{yin2012ggbio}. The bigPint package allows users to maintain consistent coloring across hierarchical clusters and when working between various plots.

\subsubsection*{Selection and aggregation}

Some techniques that are effective in data exploration may lose their efficiency and eventually fail as data size increases. Two main approaches to solving this problem are data selection and data aggregation \cite{andrienko2005blending}. Data selection means that only a subset of the full data is displayed at a given time. The data subset can be selected through queries and interactive controls which allow the user to quickly examine different data subsets \cite{andrienko2005blending}. Data aggregation means that the full dataset is divided into data subsets (called aggregates) that reduce the amount of data being simultaneously visualized. Users with large datasets should ideally be able to perform both data selection and data aggregation \cite{andrienko2005blending}. The bigPint package allows users to easily perform data selection using queries (such as thresholds and sliders) and interactive controls (such as zooming, box and lasso selection, and panning) and to perform data aggregation using hierarchical clustering. 

\subsubsection*{Shiny interactivity}

Interactive plots in the bigPint package open as Shiny applications that consist of simple dashboards with ``About'' tabs that explain how to use the applications. They also include ``Application'' tabs that provide several input fields for the user to tailor their plots. Some of these input fields are generated dynamically from the inputted dataset so that users have more convenience in how they select data subsets. In these applications, users can also download lists of selected genes and static images of interactive graphics to their local computers. 

Shiny applications can be launched on a local personal computer, hosted on a local or cloud-based server, or hosted for free on the shinyapps.io website. As such, interactive bigPint packages can be deployed on a personal computer using only a local file containing the data, the bigPint package and its dependencies, R / RStudio, and a browser recommended by Shiny (Google Chrome or Mozilla Firefox). This method does not require internet connectivity, which can be useful for users who are protecting sensitive data, analyzing or presenting data in contexts without reliable connectivity, or testing and developing applications.

% implications of the findings, limitations, practical/operational issues
\section*{Discussion}

Researchers benefit when they are able to view multiple perspective of their data, especially when working with large datasets \cite{swayne2003ggobi, cook2007interactive}. The ability to select and aggregate data, threshold data to create subsets, link between multiple plots, interact with plots, and tailor various aesthetics in intelligent ways are all useful features of the bigPint package \cite{o2010visualizing, pavlopoulos2015visualizing, kerren2007human}. We expect that bigPint will enable researchers to generate and interact with intuitive, high quality, and reproducible plots from increasingly large biological datasets. 

% main conclusions, importance and relevance of study to field 
\section*{Conclusion}

Despite the growing appreciation of the inherent value in interactive graphics, the availability of easy-to-use and effective interactive exploratory visualization tools for RNA-seq data remains limited. In this paper, we introduced new visualization tools that enable independent layers of interactive capabilities for the foreground and background of plots. We believe this methodology represents a fairly novel contribution to the field of interactive data visualization. Advocating state-of-the-art visualization tools is crucial for biology researchers to analyze and present their data and for visualization researchers to develop novel methods. Lessons learned from our open-source work may encourage the development of additional interactive visualization tools for various computational biology tasks.

% software tool requirements
\section*{Methods}

bigPint was released under the GPL-3 license. Most bigPint visualization methods were constructed using htmlwidgets \cite{htmlwidgets}, ggplot2 \cite{ggplot2}, shiny \cite{shiny}, shinyapps.io \cite{rstudio}, and plotly \cite{plotly}. bigPint methods were tested on numerous RNA-seq datasets \cite{rutter1, rutter2}. The package website was constructed using the pkgdown software \cite{pkgdown}. bigPint can be downloaded from the Bioconductor website \cite{gentleman2004bioconductor}. 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Competing interests}
  The authors declare that they have no competing interests.

\section*{Author's contributions}
    Text for this section \ldots

\section*{Acknowledgements}
  Text for this section \ldots
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file (bmc-mathphys, vancouver, spbasic).
\bibliography{bmc_article}      % Bibliography file (usually '*.bib' )
% for author-year bibliography (bmc-mathphys or spbasic)
% a) write to bib file (bmc-mathphys only)
% @settings{label, options="nameyear"}
% b) uncomment next line
%\nocite{label}

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% Do not use \listoffigures as most will included as separate files

\section*{Figures}
\begin{figure}[h!]
\caption{\csentence{Independent interactive layers of scatterplot matrix.}
A) User hovers over background hexagon to determine it contains two genes. B) User clicks on background hexagon to overlay the two corresponding genes as orange points in the foreground layer of each scatterplot. The computationally-expensive background layer of hexagons does not need to be redrawn. C) The background layer of hexagons remains interactive and the user can still hover over another hexagon of interest to determine it contains 40 genes. D) User clicks on background hexagon to overlay the 40 corresponding genes as orange points in the foreground layer of each scatterplot. This step does not require the computationally-expensive background layer of hexagons to be redrawn. Note: This figure only focused on the independent nature of the two interactive layers. Interactive scatterplot matrices in bigPint have several more useful features. Please see Table~\ref{table:table2} for more details (video, pseudocode, code, and application link).}
\label{figureSM}
\end{figure}

\begin{figure}[h!]
\caption{\csentence{Independent interactive layers of litre plot.}
A) User uses Shiny buttons to specify treatment pairs (N and P) and hexagon size (10) for drawing background hexagon layer. User can hover over hexagon of interest to determine it contains 19 genes. B) User uses Shiny buttons to specify metric (FDR) and metric order (Increasing) to establish the order in which genes will be overlaid as pink points in the foreground layer. User clicks “plot gene” button and the gene with the lowest FDR value (Glyma.19G168700.Wm82.a2.v1) is overlaid. The background layer of hexagons does not need to be redrawn. C) User clicks “plot gene” button again and the gene with the second-lowest FDR value (Glyma.13G293500.Wm82.a2.v1) is overlaid. This step does not require the background layer of hexagons to be redrawn. D) User can zoom and pan on the layers using the Plotly Modebar. Note: This figure only focused on the independent nature of the two interactive layers. Interactive litre plots in bigPint have several more useful features. Please see Table~\ref{table:table2} for more details (video, pseudocode, code, and application link).}
\label{figureLitre}
\end{figure}

\begin{figure}[h!]
\caption{\csentence{Independent interactive layers of volcano plot.}
A) User uses Shiny buttons to specify treatment pairs (N and P) and hexagon size (9) for drawing background hexagon layer. User can hover over hexagon of interest to determine it contains 1 gene. B) User uses Shiny buttons to specify log fold change and p-value thresholds. User clicks “plot gene subset” button and the subset of genes that pass the thresholds are overlaid in the foreground layer as pink points. The background layer of hexagons does not need to be redrawn. User hovers over foreground point to view gene name (Glyma.19G168700.Wm82.a2.v1). C) User uses Shiny buttons to decrease point size from 8 to 6. Foreground layer of pink points are reduced in size and the background layer of hexagons does not need to be redrawn. D) User uses Shiny buttons to update threshold values and again presses “plot gene subset” button. The subset of genes that pass the new thresholds are overlaid in the foreground layer as pink points. The background layer of hexagons does not need to be redrawn. Note: This figure only focused on the independent nature of the two interactive layers. Interactive volcano plots in bigPint have several more useful features. Please see Table~\ref{table:table2} for more details (video, pseudocode, code, and application link).}
\label{figureVolcano}
\end{figure}

\begin{figure}[h!]
\caption{\csentence{Consecutive box selection in parallel coordinate plot.}
A) User selects the Box Select tool from the Plotly Modebar. B) User specifies box selection by drawing a rectangular query. Only the genes (pink lines) inside the specified rectangle remain. C) User can hover over a gene of interest (pink line) to view its name (Glyma.11G216300.Wm82.a2.v1). D) User can zoom and pan on the plot using the Plotly Modebar. Note: This figure only focused on the consecutive box selection feature. Interactive parallel coordinate plots in bigPint have several more useful features. Please see Table~\ref{table:table2} for more details (video, pseudocode, code, and application link).}
\label{figurePCP}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
\section*{Tables}

\begin{table}[h]
\caption{Examples of independent layers of interactivity}
\begin{center}
\centering\begin{tabular}{|p{0.8cm}|p{1.2cm}|p{5cm}|p{5cm}|}
\hline
\textbf{Plot} & \textbf{Layer} & \textbf{Geom-drawing interactivity} & \textbf{Geom-manipulation interactivity} \\ \hline
\parbox[t]{2mm}{\multirow{6}{*}{\rotatebox[origin=c]{90}{\makecell{Scatterplot \\ matrix}}}}
& \multicolumn{1}{c|}{\multirow{2}{*}{Background}}
& None & User hovers over background hexagons to view gene counts \\ \cline{2-4}
& \multirow{4}{*}{Foreground} & User clicks on background hexagon to draw corresponding genes as foreground points. Background layer does not need to be redrawn & User hovers over foreground points to view gene names \\ \hline
\parbox[t]{2mm}{\multirow{7}{*}{\rotatebox[origin=c]{90}{\makecell{Litre \\ plot}}}}
& \multirow{3}{*}{Background}
& User uses Shiny buttons to specify treatment pairs and hexagon sizes for drawing background hexagons & User hovers over background hexagons to view gene counts \\ \cline{2-4}
& \multirow{4}{*}{Foreground} & User uses Shiny buttons to specify metric, metric order, and point size for drawing foreground points. Background layer does not need to be redrawn &  User hovers over foreground points to view gene names \\ \hline
\parbox[t]{2mm}{\multirow{7}{*}{\centering\rotatebox[origin=c]{90}{\makecell{Volcano \\ plot}}}}
& \multirow{3}{*}{Background}
& User uses Shiny buttons to specify treatment pairs and hexagon sizes for drawing background hexagons & User hovers over background hexagons to view gene counts \\ \cline{2-4}
& \multirow{4}{*}{Foreground} & User uses Shiny buttons to specify point size, log fold changes, p-values to draw foreground points. Background layer does not need to be redrawn & User hovers over foreground points to view gene names \\ \hline
\end{tabular}
\end{center}
\label{table:table1}
\end{table}

\begin{table}[h]
\caption{Helpful resources about interactive graphics in bigPint}
      \begin{tabular}{llllll}
        \hline
        Plot & Figure & Explanation video & Interactive application & Pseudocode & Code\\ \hline
        Scatterplot matrix & Figure~\ref{figureSM} & Link a1 & bit.ly/smplotApp & Algorithm~\ref{algorithmSM} & Link a3\\
        Litre plot & Figure~\ref{figureLitre} & Link b1 & bit.ly/litreApp & Algorithm~\ref{algorithmLitre} & Link b3\\
        Volcano plot & Figure~\ref{figureVolcano} & Link c1 & bit.ly/volcanoApp & Figure c2 & Link c3\\
        Parallel coordinate plot & Figure~\ref{figurePCP} & Link d1 & bit.ly/pcplotApp & Figure d2 & Link d3\\ \hline
      \end{tabular}
      \label{table:table2}
\end{table}











\clearpage

\DontPrintSemicolon
\begin{algorithm}[H]
\SetInd{0.2em}{1.3em}

\KwData{Data frame input by user}
\KwResult{Interactive scatterplot matrix}

/* \textit{Declare Shiny server}\;
\SetKwBlock{Fna}{\textnormal{server $\leftarrow$ function(input, output, session)\{ }}{}
\Fna{
\BlankLine
/* \textit{Declare Shiny output scatterplot matrix}\;
\SetKwBlock{Fnb}{\textnormal{output\$scatMatPlot $\leftarrow$ renderPlotly(\{ }}{}
\Fnb{
\BlankLine
/* \textit{Draw hexagons and x=y line in bottom-left corner of matrix}\;
my\_fn $\leftarrow$ function(data, mapping)\{\}\;
\BlankLine
/* \textit{Create static scatterplot matrix}\;
p $\leftarrow$ ggpairs(data, lower = list(continuous = my\_fn))\;
\BlankLine
/* \textit{Convert ggplot2::ggplot() object to plotly object}\;
ggP $\leftarrow$ ggplotly(p)\;
\BlankLine
/* \textit{Tailor plotly scatterplot matrix interactivity with JavaScript}\;
\SetKwBlock{Fnc}{\textnormal{ggPR $\leftarrow$ ggP \%$>$\% onRender("function(el, x, data)\{ }}{}
\Fnc{
\BlankLine
/* \textit{If the user clicks on the plotly scatterplot matrix object}\;
\SetKwBlock{Fnd}{\textnormal{el.on('plotly\_click', function(e)\{ }}{}
\Fnd{
\BlankLine
/* \textit{Delete any old superimposed plotly geoms (orange dots)}\;
if (x.data.length $>$ 0)\{Plotly.deleteTraces(el.id)\}\;
\BlankLine
/* \textit{Determine gene IDs selected by user click. Save as object called selID with handle called 'selID' so it can be read outside current JavaScript function back in Shiny}\;
Shiny.onInputChange('selID', selID)\;
\BlankLine
/* \textit{Create traces for selected gene IDs as orange points that state gene names upon hovering}\;
trace = \{mode: 'markers', color: 'orange', size: 6, text: selID, hoverinfo: 'text'\}\;
\BlankLine
/* \textit{Superimpose traces onto the plotly scatterplot matrix object}\;
Plotly.addTraces(el.id, Traces)\;
}\})\;
}\})\;
\BlankLine
/* \textit{Pass the R data object into the JavaScript function}\;
", data = data\;
}\})\;
\BlankLine
/* \textit{Read into Shiny the gene IDs that user clicked on}\;
selID $\leftarrow$ reactive(input\$selID)\;
\BlankLine
/* \textit{Create data subset (read counts) for only the selected gene IDs}\;
pcpDat $\leftarrow$ reactive(data[which(data\$ID \%$>$\% selID()), ])\;
\BlankLine
/* \textit{Create static box plot of the full dataset}\;
BP $\leftarrow$ ggplot(data) $+$ geom\_boxplot()\;
\BlankLine
/* \textit{Render boxplot interactive as a plotly object}\;
ggBP $\leftarrow$ ggplotly(BP)\;
\BlankLine
/* \textit{Declare Shiny output boxplot}\;
\SetKwBlock{Fne}{\textnormal{output\$boxPlot $\leftarrow$ renderPlotly(\{ }}{}
\Fne{
\BlankLine
/* \textit{Tailor interactivity of the plotly boxplot object using custom JavaScript}\;
ggBP \%$>$\% onRender("function(el, x, data)\{\;
\BlankLine
/* \textit{Create traces for selected gene IDs as orange lines that state gene names upon hovering}\;
trace = \{mode: 'lines', color: 'orange', width: 1.5, text: selID, hoverinfo: 'text'\}\;
\BlankLine
/* \textit{Push traces to be superimposed onto the plotly scatterplot matrix object}\;
Plotly.addTraces(el.id, Traces)\;
\BlankLine
/* \textit{Pass R objects into the JavaScript function}\;
\}", data = list(pcpDat = pcpDat())\;
}\})\;
}\}\;
\caption{Pseudocode for interactive scatterplot matrix}
\label{algorithmSM}
\end{algorithm}







\clearpage


\begin{algorithm}[H]
\SetInd{0.2em}{1.3em}

\KwData{Data frame input by user}
\KwResult{Interactive litre plot}

/* \textit{Declare Shiny server}\;
\SetKwBlock{Fna}{\textnormal{server $\leftarrow$ function(input, output, session)\{ }}{}
\Fna{
\BlankLine
/* \textit{User input options}\;
observeEvent(input\$goButton, values\$x $\leftarrow$ values\$x $+$ 1)\;
observeEvent(input\$selPair, values\$x $\leftarrow$ 0)\;
observeEvent(input\$selMetric, values\$x $\leftarrow$ 0)\;
observeEvent(input\$selOrder, values\$x $\leftarrow$ 0)\;
observeEvent(input\$binSize, values\$x $\leftarrow$ 0)\;
\BlankLine
/* \textit{Create reactive expression of plotly background litre plot}\;
gP $\leftarrow$ reactive({p $\leftarrow$ ggplot(data); gP $\leftarrow$ ggplotly(p)})\;
\BlankLine
/* \textit{Declare shiny output litre plot}\;
\SetKwBlock{Fnb}{\textnormal{output\$hexPlot $\leftarrow$ renderPlotly(\{ }}{}
\Fnb{
\BlankLine
/* \textit{Create reactive expression of plotly background litre plot}\;
plotlyHex $\leftarrow$ reactive(gP())\;
\BlankLine
/* \textit{Tailor interactivity of the plotly litre plot object using custom JavaScript}\;
\SetKwBlock{Fnc}{\textnormal{plotlyHex() \%$>$\% onRender("function(el, x, data)\{ }}{}
\Fnc{
\BlankLine
/* \textit{Read handle called 'points' to obtain variables sent from R into JavaScript}\;
\SetKwBlock{Fnd}{\textnormal{Shiny.addCustomMessageHandler('points', function(drawPoints)\{ }}{}
\Fnd{
\BlankLine
/* \textit{Delete any old superimposed plotly geoms (dots)}\;
if (x.data.length $>$ 0)\{Plotly.deleteTraces(el.id)\}\;
\BlankLine
/* \textit{Create traces for selected gene IDs as points that state gene names upon hovering}\;
trace = \{x: drawPoints.geneX, y: drawPoints.geneY, mode: 'markers', color: drawPoints.pointColor, size: drawPoints.pointSize, text: drawPoints.geneID, hoverinfo: 'text'\}\;
\BlankLine
/* \textit{Superimpose traces onto the plotly litre plot object}\;
Plotly.addTraces(el.id, trace)\;

}\})\; %Fnd
}\}")\; %Fnc
}\})\; %Fnb

\BlankLine
/* \textit{If the user changes the superimposed gene}\;
\SetKwBlock{Fob}{\textnormal{observe(\{ }}{}
\Fob{
\BlankLine
/* \textit{Save information about superimposed gene selected by user with a handle called 'points'. These values can then be sent from R to JavaScript}\;
session\$sendCustomMessage(type = "points", message=list(geneX=geneX, geneY=geneY, pointSize = pointSize, geneID=geneID, pointColor=pointColor))\;
}\})\; %Fob

\BlankLine
/* \textit{Declare Shiny output boxplot}\;
\SetKwBlock{Fne}{\textnormal{output\$boxPlot $\leftarrow$ renderPlotly(\{ }}{}
\Fne{
\BlankLine
/* \textit{Create reactive expression of plotly background boxplot}
BP $\leftarrow$ reactive(ggplot() $+$ geom\_boxplot())
ggBP $\leftarrow$ reactive(ggplotly(BP()))
\BlankLine
/* \textit{If the user changes the superimposed gene}\;
\SetKwBlock{Fobs}{\textnormal{observe(\{ }}{}
\Fobs{
\BlankLine
/* \textit{Save information about superimposed gene selected by user with a handle called 'lines'. These values can then be sent from R to JavaScript}\;
session\$sendCustomMessage(type = "lines", message=list(geneInfo=currGene(), geneID=geneID, pointColor=pointColor))\;
}\})\; %Fobs

\BlankLine
/* \textit{Tailor interactivity of the plotly boxplot object using custom JavaScript}\;
\SetKwBlock{Fnf}{\textnormal{ggBP() \%$>$\% onRender("function(el, x, data)\{ }}{}
\Fnf{
\BlankLine
/* \textit{Read handle called 'lines' to obtain variables sent from R into JavaScript}\;
\SetKwBlock{Fng}{\textnormal{Shiny.addCustomMessageHandler('lines', function(drawLines)\{ }}{}
\Fng{
\BlankLine
/* \textit{Delete any old superimposed plotly geoms (lines)}\;
Plotly.deleteTraces(el.id, traceLine)\;
\BlankLine
/* \textit{Create traces for selected gene IDs as lines that state gene names upon hovering}\;
traceLine = \{x: drawLines.geneInfo, y: drawLines.geneInfo, mode: 'lines', color: drawLines.pointColor, width: 2, opacity: 0.9; text: drawLines.geneID, hoverinfo: 'text'\}\;
\BlankLine
/* \textit{Superimpose traces onto the plotly litre plot object}\;
Plotly.addTraces(el.id, traceLine)\;

}\})\; %Fng
}\})\; %Fnf
}\})\; %Fne
}\}\; %Fna

\caption{Pseudocode for interactive litre plot}
\label{algorithmLitre}
\end{algorithm}












\clearpage




















\begin{algorithm}[H]
\SetInd{0.2em}{1.3em}

\KwData{Data frame input by user}
\KwResult{Interactive volcano plot}

/* \textit{Declare Shiny server}\;
\SetKwBlock{Fna}{\textnormal{server $\leftarrow$ function(input, output, session)\{ }}{}
\Fna{
\BlankLine
/* \textit{User input options}\;
observeEvent(input\$goButton, values\$x $\leftarrow$ values\$x $+$ 1)\;
observeEvent(input\$selPair, values\$x $\leftarrow$ 0)\;
observeEvent(input\$selMetric, values\$x $\leftarrow$ 0)\;
observeEvent(input\$selOrder, values\$x $\leftarrow$ 0)\;
observeEvent(input\$binSize, values\$x $\leftarrow$ 0)\;
observeEvent(input\$selPair, values\$selPair $\leftarrow$ input\$selPair)\;
\BlankLine
/* \textit{Define largest fold change dynamically based on data}\;
fcInMax $\leftarrow$ max(ldply(dataMetrics, rbind)[["logFC"]])
\BlankLine
/* \textit{Construct dynamic input Shiny slider for fold change}\;
output\$slider $\leftarrow$ renderUI({sliderInput("logFC", "Log fold change:", min=0, max=fcInMax, step=0.1)})
\BlankLine
/* \textit{Declare shiny output volcano plot}\;
\SetKwBlock{Fnb}{\textnormal{output\$volPlot $\leftarrow$ renderPlotly(\{ }}{}
\Fnb{
\BlankLine
/* \textit{Create reactive expression of plotly background volcano plot}\;
gP $\leftarrow$ reactive({p $\leftarrow$ ggplot(data); gP $\leftarrow$ ggplotly(p)})\;
\BlankLine
/* \textit{Create reactive expression of plotly background volcano plot}\;
plotlyVol $\leftarrow$ reactive(gP())\;
\BlankLine
/* \textit{Tailor interactivity of the plotly volcano plot object using custom JavaScript}\;
\SetKwBlock{Fnc}{\textnormal{plotlyVol() \%$>$\% onRender("function(el, x, data)\{ }}{}
\Fnc{
\BlankLine
/* \textit{Read handle called 'points' to obtain variables sent from R into JavaScript}\;
\SetKwBlock{Fnd}{\textnormal{Shiny.addCustomMessageHandler('points', function(drawPoints)\{ }}{}
\Fnd{
\BlankLine
/* \textit{Delete any old superimposed plotly geoms (dots)}\;
if (x.data.length $>$ 0)\{Plotly.deleteTraces(el.id)\}\;
\BlankLine
/* \textit{Create traces for selected gene IDs as points that state gene names upon hovering}\;
trace = \{x: drawPoints.geneX, y: drawPoints.geneY, mode: 'markers', color: drawPoints.pointColor, size: drawPoints.pointSize, text: drawPoints.geneID, hoverinfo: 'text'\}\;
\BlankLine
/* \textit{Superimpose traces onto the plotly volcano plot object}\;
Plotly.addTraces(el.id, trace)\;

}\})\; %Fnd
}\}")\; %Fnc
}\})\; %Fnb
\BlankLine
/* \textit{If the user changes the superimposed gene}\;
\SetKwBlock{Fob}{\textnormal{observe(\{ }}{}
\Fob{
\BlankLine
/* \textit{Save information about superimposed gene selected by user with a handle called 'points'. These values can then be sent from R to JavaScript}\;
session\$sendCustomMessage(type = "points", message=list(geneX=geneX, geneY=geneY, pointSize = pointSize, geneID=geneID, pointColor=pointColor))\;
}\})\; %Fob

\BlankLine
/* \textit{Declare Shiny output boxplot}\;
\SetKwBlock{Fne}{\textnormal{output\$boxPlot $\leftarrow$ renderPlotly(\{ }}{}
\Fne{

\BlankLine
/* \textit{Create reactive expression of plotly background boxplot}\;
BP $\leftarrow$ reactive(ggplot() $+$ geom\_boxplot())\;
ggBP $\leftarrow$ reactive(ggplotly(BP()))\;

\BlankLine
/* \textit{Tailor interactivity of the plotly boxplot object using custom JavaScript}\;
\SetKwBlock{Fnf}{\textnormal{ggBP() \%$>$\% onRender("function(el, x, data)\{ }}{}
\Fnf{
\BlankLine
/* \textit{Read handle called 'lines' to obtain variables sent from R into JavaScript}\;
\SetKwBlock{Fng}{\textnormal{Shiny.addCustomMessageHandler('lines', function(drawLines)\{ }}{}
\Fng{
\BlankLine
/* \textit{Delete any old superimposed plotly geoms (lines)}\;
Plotly.deleteTraces(el.id, traceLine)\;
\BlankLine
/* \textit{Create xArr and yArr, array of x and y values for superimposed lines, from drawLines}\;
\BlankLine
/* \textit{Create traces for selected gene IDs as lines that state gene names upon hovering}\;
traceLine = \{x: xArr, y: yArr, mode: 'lines', color: drawLines.pointColor, width: 2, opacity: 0.9; text: drawLines.geneID, hoverinfo: 'text'\}\;
\BlankLine
/* \textit{Superimpose traces onto the plotly volcano plot object}\;
Plotly.addTraces(el.id, traceLine)\;
}\})\; %Fng
}\})\; %Fnf
}\})\; %Fne

\BlankLine
/* \textit{If the user changes the superimposed gene}\;
\SetKwBlock{Fobs}{\textnormal{observe(\{ }}{}
\Fobs{
\BlankLine
/* \textit{Save information about superimposed gene selected by user with a handle called 'points'. These values can then be sent from R to JavaScript}\;
session\$sendCustomMessage(type = "lines", message=list(geneID=geneID, pointColor=pointColor))\;
}\})\; %Fob2

}\}\; %Fna
\caption{Pseudocode for interactive volcano plot}
\end{algorithm}















\clearpage
















%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% \section*{Additional Files}
%   \subsection*{Additional file 1 --- Sample additional file title}
%     Additional file descriptions text (including details of how to
%     view the file, if it is in a non-standard format or the file extension).  This might
%     refer to a multi-page table or a figure.
% 
%   \subsection*{Additional file 2 --- Sample additional file title}
%     Additional file descriptions text.


\end{backmatter}
\end{document}
